### PR TITLE
First and Last Aggregator

### PR DESCRIPTION
**Pull request by jon-wei**
Reopening https://github.com/druid-io/druid/pull/3226 from @acslk to finish review/merge

### CODE REVIEW

**Comment by fjy** [comment]
Still 👍  from me after we resolve merge conflicts

**Comment by jon-wei** [comment]
rebased

**Comment by fjy** [comment]
@nishantmonu51 can u review?

File: docs/content/querying/aggregations.md [hunk 1]
```diff
@@ -76,6 +76,58 @@ Computes the sum of values as 64-bit floating point value. Similar to `longSum`
 { "type" : "longMax", "name" : <output_name>, "fieldName" : <metric_name> }
 ```
 
+### First / Last aggregator
+
+First and Last aggregator cannot be used in ingestion spec, and should only be specified as part of queries.
```
Comment (nishantmonu51) [inline docs/content/querying/aggregations.md]: can we also document that if the rows are rolled up, the first/last value will be a the first/last value in druid row and not the raw data being ingested.

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java [hunk 2]
```diff
@@ -171,7 +171,7 @@ public AggregatorFactory getMergingFactory(AggregatorFactory other) throws Aggre
               @Override
               public AggregatorFactory apply(String input)
               {
-                return new JavaScriptAggregatorFactory(input, fieldNames, fnAggregate, fnReset, fnCombine, config);
+                return new JavaScriptAggregatorFactory(input, Lists.newArrayList(input), fnCombine, fnReset, fnCombine, config);
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: seems unrelated to this PR ?

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import io.druid.segment.LongColumnSelector;
+
+public class DoubleFirstAggregator implements Aggregator
+{
+
+  private final FloatColumnSelector valueSelector;
+  private final LongColumnSelector timeSelector;
+  private final String name;
+
+  long firstTime;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java]: make private ?

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  double firstValue;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java]: private?

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregator.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+public class LongFirstAggregator implements Aggregator
+{
+
+  private final LongColumnSelector valueSelector;
+  private final LongColumnSelector timeSelector;
+  private final String name;
+
+  long firstTime;
+  long firstValue;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregator.java]: private ?

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new LongFirstAggregator(
+        name, metricFactory.makeLongColumnSelector(fieldName),
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: minor nit: can we change the order of args, pass time column selector first as in LongFirstBufferAggregator. different orders can cause confusion.

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new DoubleFirstAggregator(
+        name, metricFactory.makeFloatColumnSelector(fieldName),
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: minor nit: can we pass time column selector before field selector to be consistent with buffered aggregator ?

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregator.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+public class DoubleLastAggregator implements Aggregator
+{
+
+  private final FloatColumnSelector valueSelector;
+  private final LongColumnSelector timeSelector;
+  private final String name;
+
+  long lastTime;
+  double lastValue;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregator.java]: private ?

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new DoubleLastAggregator(
+        name, metricFactory.makeFloatColumnSelector(fieldName),
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: minor nit: can we pass time column selector before field selector to be consistent with buffered aggregator ?

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+  {
+    int result = name.hashCode();
+    result = 31 * result + fieldName.hashCode();
+    return result;
+  }
+
+  @Override
+  public String toString()
+  {
+    return "DoubleFirstAggregatorFactory{" +
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: DoubleLast ?

File: processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregator.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+import io.druid.segment.LongColumnSelector;
+
+public class LongLastAggregator implements Aggregator
+{
+  private final LongColumnSelector valueSelector;
+  private final LongColumnSelector timeSelector;
+  private final String name;
+
+  long lastTime;
+  long lastValue;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregator.java]: private.

File: processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+    this.name = name;
+    this.fieldName = fieldName;
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new LongLastAggregator(
+        name, metricFactory.makeLongColumnSelector(fieldName),
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME)
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java]: minor nit: can we pass time column selector before field selector to be consistent with buffered aggregator ?

**Comment by nishantmonu51** [comment]
Also, would be nice to add these aggregators to existing integration tests, we can add them to ITTwitterQueryTest and ITWikipediaQueryTest

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java (hunk 2 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: it's needed due to the changes within groupby in the PR

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java (hunk 3 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregator.java]: made this and others protected, it's accessed by the corresponding AggregatorFactory in getCombiningFactory() when overriding factorize()

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java (hunk 6 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: changed order here and in similar places

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java (hunk 10 above)
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: fixed this name

**Comment by jon-wei** [comment]
@nishantmonu51 Addressed your comments aside from the one about integration tests, haven't gotten around to that yet

**Comment by jon-wei** [comment]
@nishantmonu51 @fjy 

I added some first/last aggs to ITWikipediaQueryTest

I didn't add them to ITTwitterQueryTest since I saw the queries there weren't testing the full set of aggs, only using doubleSums generally.

I noticed that some other agg types weren't represented in the integration tests (like the max/min or javascript), and it'd be convenient to have a reference to the raw data for `twitterstream` and `wikipedia_editstream`, so I'll open issues for those points

**Comment by fjy** [comment]
👍 after travis

**Comment by jon-wei** [comment]
Opened two new issues re: integration tests

https://github.com/druid-io/druid/issues/3783
https://github.com/druid-io/druid/issues/3784
